core/src/main/scala/cats/syntax/semigroup.scala

```diff
@@ -1,14 +1,16 @@
 package cats
 package syntax
 
+import cats.macros.Ops
+
 trait SemigroupSyntax {
   // TODO: use simulacrum instances eventually
   implicit def semigroupSyntax[A: Semigroup](a: A): SemigroupOps[A] =
     new SemigroupOps[A](a)
 }
 
-class SemigroupOps[A](lhs: A)(implicit A: Semigroup[A]) {
-  def |+|(rhs: A): A = A.combine(lhs, rhs)
-  def combine(rhs: A): A = A.combine(lhs, rhs)
-  def combineN(rhs: Int): A = A.combineN(lhs, rhs)
+class SemigroupOps[A: Semigroup](lhs: A) {
+  def |+|(rhs: A): A = macro Ops.binop[A, A]
```

Comment:
Nothing -- I'd be happy to add this everywhere. I wanted to add it here since these are the places where the cost of the actual work (e.g. integer addition) is so small that the overhead would be particularly unfortunate (although we will still do better than Scalaz where we would be allocating a by-name parameter every time).